indexing-service/src/main/java/io/druid/indexing/common/task/SinglePhaseParallelIndexTaskRunner.java

```diff
@@ -0,0 +1,624 @@
+  }
+
+  @GET
+  @Path("/subtaskspec/{id}")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSubTaskSpec(@PathParam("id") String id, @Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
+    if (taskMonitor != null) {
+      // Running tasks should be checked first because, in taskMonitor, subTaskSpecs are removed from runningTasks after
+      // adding them to taskHistory.
+      final MonitorEntry monitorEntry = taskMonitor.getRunningTaskMonitorEntory(id);
+      final TaskHistory<ParallelIndexSubTask> taskHistory = taskMonitor.getCompleteSubTaskSpecHistory(id);
+      final SubTaskSpec<ParallelIndexSubTask> subTaskSpec;
+
+      if (monitorEntry != null) {
+        subTaskSpec = monitorEntry.getSpec();
+      } else {
+        if (taskHistory != null) {
+          subTaskSpec = taskHistory.getSpec();
+        } else {
+          subTaskSpec = null;
+        }
+      }
+
+      if (subTaskSpec == null) {
+        return Response.status(Response.Status.NOT_FOUND).build();
+      } else {
+        return Response.ok(subTaskSpec).build();
+      }
+    } else {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+  }
+
+  @GET
+  @Path("/subtaskspec/{id}/state")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSubTaskState(@PathParam("id") String id, @Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
+    if (taskMonitor == null) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    } else {
+      // Running tasks should be checked first because, in taskMonitor, subTaskSpecs are removed from runningTasks after
+      // adding them to taskHistory.
+      final MonitorEntry monitorEntry = taskMonitor.getRunningTaskMonitorEntory(id);
+      final TaskHistory<ParallelIndexSubTask> taskHistory = taskMonitor.getCompleteSubTaskSpecHistory(id);
+
+      final SubTaskStateResponse subTaskStateResponse;
+
+      if (monitorEntry != null) {
+        subTaskStateResponse = new SubTaskStateResponse(
+            (ParallelIndexSubTaskSpec) monitorEntry.getSpec(),
+            monitorEntry.getRunningStatus(),
+            monitorEntry.getTaskHistory()
+        );
+      } else {
+        if (taskHistory != null && !taskHistory.isEmpty()) {
+          subTaskStateResponse = new SubTaskStateResponse(
+              (ParallelIndexSubTaskSpec) taskHistory.getSpec(),
+              null,
+              taskHistory.getAttemptHistory()
+          );
+        } else {
+          subTaskStateResponse = null;
+        }
+      }
+
+      if (subTaskStateResponse == null) {
+        return Response.status(Response.Status.NOT_FOUND).build();
+      } else {
+        return Response.ok(subTaskStateResponse).build();
+      }
+    }
+  }
+
+  @GET
+  @Path("/subtaskspec/{id}/history")
```

Comment:
All this data will evaporate once the task finishes right? is there any way to fetch this information after a task has finished?

Reply:
We don't have such a system yet, but yeah it's definitely needed. I think it should be stored in metastore.